### ISSUE TITLE
bcachefs: EOL kernel, out of date, unstable.

### ISSUE THREAD

**Issue by YellowOnion** [NixOS/nixpkgs#121767]
###### Checklist

This is a thread to raise discussion on how to proceed with maintenance with this package.

I've experienced several file system corruptions.

The kernel that this build pulls in is end of life, and there's no other option but to use it or hack on nixpkgs, since this package is not a patch to NixOS kernels.

[On disk format](https://www.patreon.com/posts/disk-format-for-49169499) has changed and requires the use of the [compat-branch](https://evilpiepirate.org/git/bcachefs-tools.git/log/?h=compat) tools to fsck and upgrade your disk format, this is only a major issue if you run bcachefs on root, as your system will fail to boot without first upgrading the disk format. compat-branch will need to be included as a transitional tool for a while (1-2 years?).

My proposal:

* Adapt to a patch on NixOS kernel, this way we can get the latest version of the kernel and not be stuck on x.x.0

* Remove hard dependency on specific kernel package, so users can select what kernel package they want to use.

* package both compat and master branches of bcachefs-tools.

* Use evilpiepirate git source because it's more up to date.

***

- [x] Checked the [nixpkgs master branch](https://github.com/NixOS/nixpkgs)

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

###### Project name
`nix search` name: bcachefs-tools, linuxPackages_testing_bcachefs

current version: 5.9.0 kernel, 2020.11.20
desired version: flexible kernel, open to suggestions on maintaining

###### Notify maintainers

maintainers: @davidak @chiiruno

***

https://discourse.nixos.org/t/systemd-random-seed-service-constantly-fails/12137

@hyperfekt made this [script here](https://raw.githubusercontent.com/hyperfekt/nix-exp/master/config/bcachefs-support.nix) which we can probably adapt to be a more robust solution.

**Comment by hyperfekt** [comment]
Frankly if it was my choice I would have removed bcachefs support from NixOS and instead maintained a wiki page with exact instructions on how to set it up. Having it available normally implies a level of stability that simply isn't provided by upstream; and I don't see how someone would run bcachefs without closely watching what happens upstream anyway.

**Comment by hyperfekt** [comment]
Kent Overstreet (the developer of bcachefs) has been talking about getting it more stable and implying upstreaming into the maintained Linux kernel isn't too far away, and if that's true this problem may solve itself without much intervention from our side and not worth a lot disruption right now.

The question of course is how close those things actually are, which is *very* hard to estimate.

**Comment by YellowOnion** [comment]
I think you're right, I don't think it should have been merged in to nixpkgs, especially with how much of a "hack" it is at the moment, maybe NUR would be a better place.


I wouldn't mind maintaining something on NUR but I can't even get my kernel to compile with the diff applied. And have zero clue how to inspect the failed build to figure what the problem is.

```
  CC [M]  drivers/gpu/drm/nouveau/nouveau_led.o
  LD [M]  drivers/gpu/drm/nouveau/nouveau.o
  AR      drivers/gpu/built-in.a
  AR      drivers/built-in.a
make: *** [../Makefile:185: __sub-make] Error 2
note: keeping build directory '/tmp/nix-build-linux-5.10.30.drv-0'
builder for '/nix/store/73rbks2yv3sxm4zmcpxkx9daa1bg6hkp-linux-5.10.30.drv' failed with exit code 2
```

**Comment by davidak** [comment]
@YellowOnion thanks for bringing this topic up!

>Adapt to a patch on NixOS kernel, this way we can get the latest version of the kernel and not be stuck on x.x.0

this would break it in unexpected ways and add maintenance burden since we would move away from upstream. it would break the main use case of the package (testing what upstream provides and report issues)

>Remove hard dependency on specific kernel package, so users can select what kernel package they want to use.

this does not work. you have to wait until bcachefs support is upstreamed to mainline linux

>package both compat and master branches of bcachefs-tools.

i have not looked into it lately, but sounds resonable. we also need documentation how it is supposed to be used

>Use evilpiepirate git source because it's more up to date.

maybe. github provides more efficient download i think

>Having it available normally implies a level of stability that simply isn't provided by upstream

the package is called `linuxPackages_testing_bcachefs` for a reason. of course we can't provide better software than upstream

i have planned to add documentation. that should make it more clear

it is just for testing it and reporting issues upstream until it is stable enough and get upstreamed

>I don't think it should have been merged in to nixpkgs, especially with how much of a "hack" it is at the moment

i disagree. it is not a hack. the package provides integration of the bcachefs upstream kernel at the time of it's last update. it has tests and everything

since it's just for testing, it does not have a high priority and is only updated from time to time. feel free to help!

the package makes it easy to try it out for users. this is not the case with wiki instructions or NUR. i suggest contributing to nixpkgs instead!

**Comment by symphorien** [comment]
If you end up choosing to remove the bcachefs kernel from nixpkgs, it could be nice to do so before 21.05 is released.

**Comment by YellowOnion** [comment]
> i hope you have reported them upstream, so they can get fixed.

The fixes are in master, and this is why I brought up this thread.

When you specify `boot.supportedFilesystems = [ "bcachefs" ];` it pulls in Kent's 5.9.0, which hasn't had any commits in 5 months, it's on a kernel that's even older, this is what I mean by "hardcoded", I don't know if NixOS has a way for a package to "provide" a specific feature, while another one demand it, so we can get more flexibility.

It should be easy enough to provide patches for 5.9.x, 5.10.x, and 5.11.x, with the NixOS kernel on the various branches Kent has provided.

I'm new to NixOS so just struggling to understand the plumbing of a lot of this stuff.

I did eventually get my patch to compile last night, the problem was the error happened far earlier (like a good 5mins earlier) I don't know why the kernel keeps building uselessly.

@symphorien I think this would break too much stuff for people who have come to depend on this package.

I guess I should go investigate how to contribute to nixpkgs.

**Comment by hyperfekt** [comment]
While generally the bcachefs patchset applies just fine throughout all patches of a minor (in semantic version speak), it is specific to that minor and usually needs work to be adapted to it. Kent only maintains a singular minor on top of which bcachefs is developed. All older minors would either have a severely outdated version of bcachefs or would need rebasing work.The latter is true for minors newer than the current as well.

The current state of the bcachefs upstream however is that it is branched off from when a new minor is released, meaning it does not receive changes from newer patch versions. What I have done for myself is to apply the bcachefs patchset on top of the most recent patch version of that minor in nixpkgs and it has never not applied cleanly, or produced issues that did not also occur on the upstream version.

In conclusion I recommend that if nixpkgs continues to provide a bcachefs-enabled kernel it adopt the same method.

My problem with the bcachefs enabled kernel in nixpkgs is twofold:
1) It gets pulled in automatically by specifying bcachefs as a filesystem, which does not allow for the user to be educated on the risk they are taking in using that filesystem.
2) A newer version of the bcachefs patchset often renders filesystems unmountable, leading to serious disruption. So any updates to the version of the bcachefs patchset should be undertaken by the user, and not happen automatically simply by changing their revision of nixpkgs.

**Comment by davidak** [comment]
>When you specify boot.supportedFilesystems = [ "bcachefs" ]; it pulls in Kent's 5.9.0, which hasn't had any commits in 5 months, it's on a kernel that's even older, this is what I mean by "hardcoded",

@YellowOnion i see. unlike other distros, nixos provides tools to customize packages. so you can update the package in your config without changing nixpkgs. i'm not sure what the best method is right now, i used `nixpkgs.config.packageOverrides` last time. maybe the documentation helps you or have a look in the forum, chat or wiki https://nixos.wiki/wiki/Overlays#Examples_of_overlays

>It should be easy enough to provide patches for 5.9.x, 5.10.x, and 5.11.x, with the NixOS kernel on the various branches Kent has provided.

i don't think we have the resources to maintain 3 kernels when the one we have is that outdated ;)

>I'm new to NixOS so just struggling to understand the plumbing of a lot of this stuff.

it's actually quiet easy. i have never compiled a custom kernel and was able to package the initial package. have a look at the commits https://github.com/NixOS/nixpkgs/commits/a8758fdce5c95a1d20f599b4a6b11f862db1252f/pkgs/os-specific/linux/kernel/linux-testing-bcachefs.nix

>I guess I should go investigate how to contribute to nixpkgs.

yes. i think most users will come to the point where they become contributors, even it the contributions are small, like updating a package. i think such contributions are easier here than in other distros. just create a pull request on github

it is extremely empowering when you can just fix the issue that bothers you yourself or just update the package that wasn't touched by the maintainers for 6 months ;) everyone is allowed to improve packages, not only the official maintainers

and when you are a user of bcachefs and regularly update it, feel free to add yourself as a maintainer and help keep it up to date :)

**Comment by YellowOnion** [comment]
> i don't think we have the resources to maintain 3 kernels when the one we have is that outdated ;)

the branches for 5.9, and 5.10 are stagnant, once it works it works. We don't have to "maintain" anything, just master.

we just need automation, leavaging: https://github.com/NixOS/nixpkgs/blob/master/maintainers/scripts/update.nix

I already have patches for ever signification push Kent made here: https://github.com/YellowOnion/bcachefs-patches

**Comment by davidak** [comment]
Why would anyone want to use a bcachefs kernel that is not the latest available?

**Comment by symphorien** [comment]
> @symphorien I think this would break too much stuff for people who have come to depend on this package.

I suppose marking the kernel versions which result in data loss can be a good middle ground. The release notes can tell people to add allowBroken to their config until they update to a newer version.

**Comment by hyperfekt** [comment]
For now I've updated the patchset and made the package follow revisions of the respective minor in #123765.
Frankly I expect bcachefs to be close enough to stabilization now that most people who will be bitten by it probably already have been.

**Comment by Madouura** [comment]
AFAIK, the parts of bcachefs that need to be mainlined into the kernel are well on their way.
Doing some cleanup on bcachefs and related issues, do we still need this open?
